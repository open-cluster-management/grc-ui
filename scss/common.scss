/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/


@import 'node_modules/carbon-components/scss/globals/scss/styles';
@import 'node_modules/carbon-components/scss/globals/scss/_colors';
@import 'node_modules/carbon-components/scss/globals/scss/_helper-mixins';
@import 'header-carbon-v9-migration';
@import 'variables';
@import '_red-hat-font';

@include redhat-font-face;

html, body, div, span, applet, object, iframe, h1, h2, h3, h4, h5, h6, p, blockquote, pre, a, abbr, acronym, address, big, cite, code, del, dfn, em, img, ins, kbd, q, s, samp, small, strike, strong, sub, sup, tt, var, b, u, i, center, dl, dt, dd, ol, ul, li, fieldset, form, label, legend, table, caption, tbody, tfoot, thead, tr, th, td, article, aside, canvas, details, embed, figure, figcaption, footer, header, hgroup, menu, nav, output, ruby, section, summary, time, mark, audio, video {
  font-family: RedHatText;
}

.bx--label, .bx--search-input, .bx--overflow-menu-options__btn, .bx--pagination__text, .bx--text-input, .bx--tooltip--icon__top::before, .bx--tooltip--icon__top::after, .bx--tooltip--icon__bottom::before, .bx--tooltip--icon__bottom::after {
  font-family: RedHatText;
}

html {
  height: 100%;
}

body {
  min-height: 100%;
  height: 100%;
  text-rendering: geometricPrecision;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  .bx--data-table-v2--zebra tbody tr:nth-child(2n) {
    background-color: $color__white;
  }
  .bx--tooltip--shown {
    font-size: .875rem;
  }
}

.page-content-container {
  display: flex;
  flex-direction: column;
  padding: 1.5rem 5% 3.375rem;

  .bx--select--inline{
    svg.bx--select__arrow{
      position: absolute
    }
  }
}

.content-spinner {
  position: fixed;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}

.resource-list-filter {
  margin-top: -1.5rem;
  padding-bottom: 3rem;
}

.bx--loading-overlay {
  z-index: 10000; // need higher z-index with modals open
}

.expand-vertically {
  height: 100%;
}

.nested-expandable-row {
  display: flex;
  flex-direction: column;
  .nested-expandable-row-header {
    font-weight: bold;
    color: $color__blue-50;
  }
  .nested-expandable-row-content {
    margin-top: -2rem;
  }
}

.no-resource {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  width: 32.375rem;
  height: 25.6875rem;
  color: $color__blue-90;
  padding: 1.25rem 2.5rem;
  background-color: transparent;
  box-shadow: none;
  border: none;
  margin: 3rem auto 1rem auto;

  .button-group {
    display: flex;
    .buttons {
      margin: 0 1rem;
      button:hover {
        .doc-link {
          color: white;
        }
      }
      .doc-link {
        text-decoration: none;
      }
    }
  }

  .no-resource-icon {
    width: 22.5rem;
    height: 12.5rem;
  }

  .no-resource-title {
    font-weight: 500;
    margin-bottom: 0.5625rem;
    font-size: 1.8125rem;
    text-align: center;
    color: $grc_color_blue;
  }

  .no-resource-detail {
    font-weight: 300;
    font-size: 1rem;
    margin-bottom: 1.3125rem;
    text-align: center;
    line-height: 1.2;
    color: $color__navy-gray-6;

    > a {
      font-weight: 700;
      margin-right: 0.3125rem;
      font-size: .875rem;
    }
  }
}

.yamlEditorContainer {
  #brace-editor {
    min-width: 31.25rem;
    min-height: 31.25rem
  }
}
