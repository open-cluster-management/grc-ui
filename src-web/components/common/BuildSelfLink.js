/* Copyright (c) 2021 Red Hat, Inc. */
'use strict'
import _ from 'lodash'

export const buildSelfLinK = (data, kind) => {
  const apiGroupVersion = _.get(data, 'raw.apiVersion')
  const namespace = _.get(data, 'metadata.namespace')
  const name = _.get(data, 'metadata.name')
  const selfLink = _.get(data, 'metadata.selfLink')
  if (apiGroupVersion && namespace && kind && name) {
    return `/apis/${apiGroupVersion}/namespaces/${namespace}/${kind}/${name}`
  } else if (selfLink) {
    return selfLink
  }
  return ''
}
