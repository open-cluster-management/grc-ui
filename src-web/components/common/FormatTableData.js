/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'
import _ from 'lodash'

export const formatPoliciesToClustersTableData = (policies) => {
  const clusterTableData = []
  const clusterMap = new Map()
  if (Array.isArray(policies) && policies.length > 0) {
    policies.forEach((policy) => {
      const statuses = _.get(policy, 'raw.status.status', {})
      // step 1, build policies info grouped by each unique cluster name
      if (Array.isArray(statuses) && statuses.length > 0) {
        statuses.forEach((singleClusterStatus) => {
          const clusterName = singleClusterStatus.clustername
          if (!clusterMap.has(clusterName)) {
            clusterMap.set(clusterName, [])
          }
          let compliantNum = 0
          let nonCompliantNum = 0
          let nameSpace = '-'
          let consoleURL = '-'
          let nonCompliantPolicyName = ''
          if (policy.clusterNS && typeof policy.clusterNS === 'object' && policy.clusterNS[clusterName]) {
            nameSpace = policy.clusterNS[clusterName]
          }
          if (policy.clusterConsoleURL && typeof policy.clusterConsoleURL === 'object' && policy.clusterConsoleURL[clusterName]) {
            consoleURL = policy.clusterConsoleURL[clusterName]
          }
          const policyCompliantStatusOnCluster = _.get(singleClusterStatus, 'compliant', '').trim().toLowerCase()
          if (policyCompliantStatusOnCluster === 'compliant') {
            compliantNum = 1
          } else if (policyCompliantStatusOnCluster === 'noncompliant') {
            nonCompliantNum = 1
            nonCompliantPolicyName = _.get(policy, 'metadata.name', '-')
          } else {
            nonCompliantPolicyName = _.get(policy, 'metadata.name', '-')
          }
          const singlePolicyData = {
            nameSpace,
            compliantNum,
            nonCompliantNum,
            nonCompliantPolicyName,
            consoleURL,
          }
          clusterMap.get(clusterName).push(singlePolicyData)
        })
      }
    })

    // step 2, combine single policy info in policiesData together under each unique cluster name
    for (const [clusterName, policiesData] of clusterMap.entries()) {
      let namespace = '-'
      let consoleURL = '-'
      let totalCompliantNum = 0
      let totalNonCompliantNum = 0
      const totalPoliciesUnderCluster = policiesData.length
      const nonCompliant = []
      policiesData.forEach((singlePolicyData) => {
        namespace = singlePolicyData.nameSpace
        if (singlePolicyData.nonCompliantPolicyName) {
          nonCompliant.push(singlePolicyData.nonCompliantPolicyName)
        }
        totalCompliantNum += singlePolicyData.compliantNum
        totalNonCompliantNum += singlePolicyData.nonCompliantNum
        consoleURL = singlePolicyData.consoleURL
      })
      const singleClusterData = {
        cluster: clusterName,
        namespace,
        violation: `${totalNonCompliantNum}/${totalPoliciesUnderCluster}/${totalPoliciesUnderCluster-totalCompliantNum-totalNonCompliantNum}`,
        nonCompliant,
        consoleURL,
      }
      clusterTableData.push(singleClusterData)
    }
  }
  return clusterTableData
}

export const formatExpandablePolicies = (policies) => {
  const result = []
  if (policies) {
    policies.forEach(policy => {
      if(policy) {
        const subItems = [{name: 'policy.pb', items: policy.placementBindings ? policy.placementBindings.map(pb => pb.metadata.name) : []},
          {name: 'policy.pp', items: policy.placementPolicies ? policy.placementPolicies.map(pp => pp.metadata.name) : []}]
        result.push({...policy, subItems})
      }
    })
  }
  return result
}
