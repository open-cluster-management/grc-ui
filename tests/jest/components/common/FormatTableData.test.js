/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
/* Copyright (c) 2020 Red Hat, Inc. */
'use strict'

import { policies, policies2, policies3, findings } from './CommonTestingData'
import { formatPoliciesToClustersTableData, formatFindingsToClustersTableData, formatExpandablePolicies } from '../../../../src-web/components/common/FormatTableData'

describe('formatPoliciesToClustersTableData', () => {
  it('should correctly format', () => {
    expect(formatPoliciesToClustersTableData(policies)).toMatchSnapshot()
  })
})

describe('formatPoliciesToClustersTableData', () => {
  it('should correctly format', () => {
    expect(formatPoliciesToClustersTableData(policies2)).toMatchSnapshot()
  })
})

describe('formatPoliciesToClustersTableData', () => {
  it('should correctly format', () => {
    expect(formatPoliciesToClustersTableData(policies3)).toMatchSnapshot()
  })
})

describe('formatPoliciesToClustersTableData', () => {
  it('should correctly format', () => {
    expect(formatPoliciesToClustersTableData(null)).toMatchSnapshot()
  })
})

describe('formatFindingsToClustersTableData', () => {
  it('should correctly format', () => {
    expect(formatFindingsToClustersTableData(findings)).toMatchSnapshot()
  })
})

describe('formatFindingsToClustersTableData', () => {
  it('should correctly format', () => {
    expect(formatFindingsToClustersTableData(null)).toMatchSnapshot()
  })
})

describe('formatExpandablePolicies', () => {
  it('should correctly format', () => {
    expect(formatExpandablePolicies(policies)).toMatchSnapshot()
  })
})

describe('formatExpandablePolicies', () => {
  it('should correctly format', () => {
    expect(formatExpandablePolicies(null)).toMatchSnapshot()
  })
})

describe('formatExpandablePolicies', () => {
  it('should correctly format', () => {
    expect(formatExpandablePolicies([null, null])).toMatchSnapshot()
  })
})

