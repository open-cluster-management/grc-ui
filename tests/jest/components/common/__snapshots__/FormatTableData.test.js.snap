// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatExpandablePolicies should correctly format 1`] = `
Array [
  Object {
    "__typename": "Compliance",
    "clusterCompliant": "0/2",
    "metadata": Object {
      "__typename": "Metadata",
      "annotations": Object {
        "policy.open-cluster-management.io/categories": "SystemAndCommunicationsProtections",
        "policy.open-cluster-management.io/controls": "VA",
        "policy.open-cluster-management.io/standards": "BSA",
        "seed-generation": "1",
      },
      "name": "my-policy-test",
      "namespace": "calamari",
      "resourceVersion": "1524346",
      "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/my-policy-test",
    },
    "name": "my-policy-test",
    "namespace": "calamari",
    "placementBindings": Array [
      Object {
        "__typename": "PlacementBinding",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "binding-my-policy-test",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementbindings/binding-my-policy-test",
        },
      },
    ],
    "placementPolicies": Array [
      Object {
        "__typename": "PlacementPolicy",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "placement-my-policy-test",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementpolicies/placement-my-policy-test",
        },
      },
    ],
    "policyCompliant": "0/2",
    "raw": Object {
      "apiVersion": "policy.open-cluster-management.io/v1",
      "kind": "Policy",
      "metadata": Object {
        "annotations": Object {
          "policy.open-cluster-management.io/categories": "SystemAndCommunicationsProtections",
          "policy.open-cluster-management.io/controls": "VA",
          "policy.open-cluster-management.io/standards": "BSA",
          "seed-generation": "1",
        },
        "creationTimestamp": "2019-07-23T21:12:04Z",
        "finalizers": Array [
          "propagator.finalizer.calamari.ibm.com",
        ],
        "generation": 7,
        "name": "my-policy-test",
        "namespace": "calamari",
        "resourceVersion": "1524346",
        "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/my-policy-test",
        "uid": "85840587-ad8e-11e9-8a41-005056a061f1",
      },
      "spec": Object {
        "-templates": Array [
          Object {
            "Definition": Object {
              "apiVersion": "v1",
              "kind": "Pod",
              "metadata": Object {
                "name": "nginx-pod",
              },
              "spec": Object {
                "containers": Array [
                  Object {
                    "image": "nginx:1.7.9",
                    "name": "nginx",
                    "ports": Array [
                      Object {
                        "containerPort": 80,
                      },
                    ],
                  },
                ],
              },
            },
            "complianceType": "musthave",
            "status": Object {
              "Validity": Object {},
            },
          },
        ],
        "complianceType": "musthave",
        "namespaces": Object {
          "exclude": Array [
            "kube-*",
          ],
          "include": Array [
            "default",
          ],
        },
        "policy-templates": Array [
          Object {
            "Definition": Object {
              "apiVersion": "policies.ibm.com/v1alpha1",
              "kind": "MutationPolicy",
              "metadata": Object {
                "label": Object {
                  "category": "System-Integrity",
                },
                "name": "mutation-policy-example",
              },
              "spec": Object {
                "conditions": Object {
                  "ownership": Array [
                    "ReplicaSet",
                    "Deployment",
                    "DeamonSet",
                    "ReplicationController",
                  ],
                },
                "namespaceSelector": Object {
                  "exclude": Array [
                    "kube-system",
                  ],
                  "include": Array [
                    "default",
                    "kube-*",
                  ],
                },
                "remediationAction": "enforce",
              },
            },
            "status": Object {
              "Validity": Object {},
            },
          },
        ],
        "remediationAction": "enforce",
      },
      "status": Object {
        "placementBindings": Array [
          "binding-my-policy-test",
        ],
        "placementPolicies": Array [
          "placement-my-policy-test",
        ],
        "status": Object {
          "cluster1": Object {
            "aggregatePoliciesStatus": Object {
              "my-policy-test": Object {
                "compliant": "Compliant",
              },
            },
            "clustername": "cluster1",
            "compliant": "Compliant",
          },
          "clusterhub": Object {
            "aggregatePoliciesStatus": Object {
              "my-policy-test": Object {
                "compliant": "Compliant",
              },
            },
            "clustername": "clusterhub",
            "compliant": "Compliant",
          },
        },
      },
    },
    "remediation": "enforce",
    "subItems": Array [
      Object {
        "items": Array [
          "binding-my-policy-test",
        ],
        "name": "policy.pb",
      },
      Object {
        "items": Array [
          "placement-my-policy-test",
        ],
        "name": "policy.pp",
      },
    ],
  },
  Object {
    "__typename": "Compliance",
    "clusterCompliant": "1/2",
    "metadata": Object {
      "__typename": "Metadata",
      "annotations": Object {
        "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.open-cluster-management.io/v1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.open-cluster-management.io/categories\\":\\"SystemAndInformationIntegrity\\",\\"policy.open-cluster-management.io/controls\\":\\"MutationAdvisor\\",\\"policy.open-cluster-management.io/standards\\":\\"NIST\\"},\\"name\\":\\"policy-ma\\",\\"namespace\\":\\"calamari\\"},\\"spec\\":{\\"policy-templates\\":[{\\"Definition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"MutationPolicy\\",\\"metadata\\":{\\"label\\":{\\"category\\":\\"System-Integrity\\"},\\"name\\":\\"mutation-policy-example\\"},\\"spec\\":{\\"conditions\\":{\\"ownership\\":[\\"ReplicaSet\\",\\"Deployment\\",\\"DeamonSet\\",\\"ReplicationController\\"]},\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\",\\"kube-*\\"]},\\"remediationAction\\":\\"enforce\\"}}}],\\"remediationAction\\":\\"enforce\\"}}
",
        "policy.open-cluster-management.io/categories": "SystemAndInformationIntegrity",
        "policy.open-cluster-management.io/controls": "MutationAdvisor",
        "policy.open-cluster-management.io/standards": "NIST",
        "seed-generation": "1",
      },
      "name": "policy-ma",
      "namespace": "calamari",
      "resourceVersion": "1347579",
      "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/policy-ma",
    },
    "name": "policy-ma",
    "namespace": "calamari",
    "placementBindings": Array [
      Object {
        "__typename": "PlacementBinding",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "binding-ma",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementbindings/binding-ma",
        },
      },
    ],
    "placementPolicies": Array [
      Object {
        "__typename": "PlacementPolicy",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "placement-ma",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementpolicies/placement-ma",
        },
      },
    ],
    "policyCompliant": "1/2",
    "raw": Object {
      "apiVersion": "policy.open-cluster-management.io/v1",
      "kind": "Policy",
      "metadata": Object {
        "annotations": Object {
          "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.open-cluster-management.io/v1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.open-cluster-management.io/categories\\":\\"SystemAndInformationIntegrity\\",\\"policy.open-cluster-management.io/controls\\":\\"MutationAdvisor\\",\\"policy.open-cluster-management.io/standards\\":\\"NIST\\"},\\"name\\":\\"policy-ma\\",\\"namespace\\":\\"calamari\\"},\\"spec\\":{\\"policy-templates\\":[{\\"Definition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"MutationPolicy\\",\\"metadata\\":{\\"label\\":{\\"category\\":\\"System-Integrity\\"},\\"name\\":\\"mutation-policy-example\\"},\\"spec\\":{\\"conditions\\":{\\"ownership\\":[\\"ReplicaSet\\",\\"Deployment\\",\\"DeamonSet\\",\\"ReplicationController\\"]},\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\",\\"kube-*\\"]},\\"remediationAction\\":\\"enforce\\"}}}],\\"remediationAction\\":\\"enforce\\"}}
",
          "policy.open-cluster-management.io/categories": "SystemAndInformationIntegrity",
          "policy.open-cluster-management.io/controls": "MutationAdvisor",
          "policy.open-cluster-management.io/standards": "NIST",
          "seed-generation": "1",
        },
        "creationTimestamp": "2019-07-16T13:18:48Z",
        "finalizers": Array [
          "propagator.finalizer.calamari.ibm.com",
        ],
        "generation": 45,
        "name": "policy-ma",
        "namespace": "calamari",
        "resourceVersion": "1347579",
        "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/policy-ma",
        "uid": "3f218d45-a7cc-11e9-8a41-005056a061f1",
      },
      "spec": Object {
        "namespaces": Object {},
        "policy-templates": Array [
          Object {
            "Definition": Object {
              "apiVersion": "policies.ibm.com/v1alpha1",
              "kind": "MutationPolicy",
              "metadata": Object {
                "label": Object {
                  "category": "System-Integrity",
                },
                "name": "mutation-policy-example",
              },
              "spec": Object {
                "conditions": Object {
                  "ownership": Array [
                    "ReplicaSet",
                    "Deployment",
                    "DeamonSet",
                    "ReplicationController",
                  ],
                },
                "namespaceSelector": Object {
                  "exclude": Array [
                    "kube-system",
                  ],
                  "include": Array [
                    "default",
                    "kube-*",
                  ],
                },
                "remediationAction": "enforce",
              },
            },
            "status": Object {
              "Validity": Object {},
            },
          },
        ],
        "remediationAction": "enforce",
      },
      "status": Object {
        "placementBindings": Array [
          "binding-ma",
        ],
        "placementPolicies": Array [
          "placement-ma",
        ],
        "status": Object {
          "cluster1": Object {
            "aggregatePoliciesStatus": Object {
              "policy-ma": Object {
                "compliant": "NonCompliant",
              },
            },
            "clustername": "cluster1",
            "compliant": "NonCompliant",
          },
          "clusterhub": Object {
            "aggregatePoliciesStatus": Object {
              "policy-ma": Object {
                "compliant": "Compliant",
              },
            },
            "clustername": "clusterhub",
            "compliant": "Compliant",
          },
        },
      },
    },
    "remediation": "enforce",
    "subItems": Array [
      Object {
        "items": Array [
          "binding-ma",
        ],
        "name": "policy.pb",
      },
      Object {
        "items": Array [
          "placement-ma",
        ],
        "name": "policy.pp",
      },
    ],
  },
  Object {
    "__typename": "Compliance",
    "clusterCompliant": "0/2",
    "metadata": Object {
      "__typename": "Metadata",
      "annotations": Object {
        "policy.open-cluster-management.io/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
        "policy.open-cluster-management.io/controls": "MutationAdvisor,CertificateManager",
        "policy.open-cluster-management.io/standards": "NIST,HIPAA",
        "seed-generation": "1",
      },
      "name": "policy-pod",
      "namespace": "calamari",
      "resourceVersion": "1524369",
      "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/policy-pod",
    },
    "name": "policy-pod",
    "namespace": "calamari",
    "placementBindings": Array [
      Object {
        "__typename": "PlacementBinding",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "binding-pod",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementbindings/binding-pod",
        },
      },
    ],
    "placementPolicies": Array [
      Object {
        "__typename": "PlacementPolicy",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "placement-pod",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementpolicies/placement-pod",
        },
      },
    ],
    "policyCompliant": "0/2",
    "raw": Object {
      "apiVersion": "policy.open-cluster-management.io/v1",
      "kind": "Policy",
      "metadata": Object {
        "annotations": Object {
          "policy.open-cluster-management.io/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
          "policy.open-cluster-management.io/controls": "MutationAdvisor,CertificateManager",
          "policy.open-cluster-management.io/standards": "NIST,HIPAA",
          "seed-generation": "1",
        },
        "creationTimestamp": "2019-07-16T13:16:46Z",
        "finalizers": Array [
          "propagator.finalizer.calamari.ibm.com",
        ],
        "generation": 20,
        "name": "policy-pod",
        "namespace": "calamari",
        "resourceVersion": "1524369",
        "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/policy-pod",
        "uid": "f61a25eb-a7cb-11e9-8a41-005056a061f1",
      },
      "spec": Object {
        "-templates": Array [
          Object {
            "Definition": Object {
              "apiVersion": "v1",
              "kind": "Pod",
              "metadata": Object {
                "name": "nginx-pod",
              },
              "spec": Object {
                "containers": Array [
                  Object {
                    "image": "nginx:1.7.9",
                    "name": "nginx",
                    "ports": Array [
                      Object {
                        "containerPort": 80,
                      },
                    ],
                  },
                ],
              },
            },
            "complianceType": "musthave",
            "status": Object {
              "Validity": Object {},
            },
          },
        ],
        "complianceType": "musthave",
        "namespaces": Object {
          "exclude": Array [
            "kube*",
          ],
          "include": Array [
            "default",
          ],
        },
        "remediationAction": "inform",
      },
      "status": Object {
        "placementBindings": Array [
          "binding-pod",
        ],
        "placementPolicies": Array [
          "placement-pod",
        ],
        "status": Object {
          "cluster1": Object {
            "aggregatePoliciesStatus": Object {
              "policy-pod": Object {
                "compliant": "Compliant",
              },
            },
            "clustername": "cluster1",
            "compliant": "Compliant",
          },
          "clusterhub": Object {
            "aggregatePoliciesStatus": Object {
              "policy-pod": Object {
                "compliant": "Compliant",
              },
            },
            "clustername": "clusterhub",
            "compliant": "Compliant",
          },
        },
      },
    },
    "remediation": "inform",
    "subItems": Array [
      Object {
        "items": Array [
          "binding-pod",
        ],
        "name": "policy.pb",
      },
      Object {
        "items": Array [
          "placement-pod",
        ],
        "name": "policy.pp",
      },
    ],
  },
  Object {
    "__typename": "Compliance",
    "clusterCompliant": "0/2",
    "metadata": Object {
      "__typename": "Metadata",
      "annotations": Object {
        "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.open-cluster-management.io/v1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.open-cluster-management.io/categories\\":\\"SystemAndCommunicationsProtections,SystemAndInformationIntegrity\\",\\"policy.open-cluster-management.io/controls\\":\\"MutationAdvisor,VA\\",\\"policy.open-cluster-management.io/standards\\":\\"NIST,HIPAA\\"},\\"name\\":\\"policy-role\\",\\"namespace\\":\\"calamari\\"},\\"spec\\":{\\"namespaces\\":{\\"exclude\\":[\\"kube*\\"],\\"include\\":[\\"default\\"]},\\"remediationAction\\":\\"inform\\",\\"role-templates\\":[{\\"apiVersion\\":\\"roletemplate.calamari.ibm.com/v1alpha1\\",\\"complianceType\\":\\"musthave\\",\\"metadata\\":{\\"name\\":\\"operator-role-policy\\",\\"namespace\\":\\"\\"},\\"rules\\":[{\\"complianceType\\":\\"musthave\\",\\"policyRule\\":{\\"apiGroups\\":[\\"extensions\\",\\"apps\\"],\\"resources\\":[\\"deployments\\"],\\"verbs\\":[\\"get\\",\\"list\\",\\"watch\\",\\"create\\",\\"delete\\",\\"patch\\"]}}],\\"selector\\":{\\"matchLabels\\":{\\"dev\\":\\"true\\"}}}]}}
",
        "policy.open-cluster-management.io/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
        "policy.open-cluster-management.io/controls": "MutationAdvisor,VA",
        "policy.open-cluster-management.io/standards": "NIST,HIPAA",
        "seed-generation": "1",
      },
      "name": "policy-role",
      "namespace": "calamari",
      "resourceVersion": "1513562",
      "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/policy-role",
    },
    "name": "policy-role",
    "namespace": "calamari",
    "placementBindings": Array [
      Object {
        "__typename": "PlacementBinding",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "binding-role",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementbindings/binding-role",
        },
      },
    ],
    "placementPolicies": Array [
      Object {
        "__typename": "PlacementPolicy",
        "metadata": Object {
          "__typename": "Metadata",
          "name": "placement-role",
          "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/placementpolicies/placement-role",
        },
      },
    ],
    "policyCompliant": "0/2",
    "raw": Object {
      "apiVersion": "policy.open-cluster-management.io/v1",
      "kind": "Policy",
      "metadata": Object {
        "annotations": Object {
          "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.open-cluster-management.io/v1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.open-cluster-management.io/categories\\":\\"SystemAndCommunicationsProtections,SystemAndInformationIntegrity\\",\\"policy.open-cluster-management.io/controls\\":\\"MutationAdvisor,VA\\",\\"policy.open-cluster-management.io/standards\\":\\"NIST,HIPAA\\"},\\"name\\":\\"policy-role\\",\\"namespace\\":\\"calamari\\"},\\"spec\\":{\\"namespaces\\":{\\"exclude\\":[\\"kube*\\"],\\"include\\":[\\"default\\"]},\\"remediationAction\\":\\"inform\\",\\"role-templates\\":[{\\"apiVersion\\":\\"roletemplate.calamari.ibm.com/v1alpha1\\",\\"complianceType\\":\\"musthave\\",\\"metadata\\":{\\"name\\":\\"operator-role-policy\\",\\"namespace\\":\\"\\"},\\"rules\\":[{\\"complianceType\\":\\"musthave\\",\\"policyRule\\":{\\"apiGroups\\":[\\"extensions\\",\\"apps\\"],\\"resources\\":[\\"deployments\\"],\\"verbs\\":[\\"get\\",\\"list\\",\\"watch\\",\\"create\\",\\"delete\\",\\"patch\\"]}}],\\"selector\\":{\\"matchLabels\\":{\\"dev\\":\\"true\\"}}}]}}
",
          "policy.open-cluster-management.io/categories": "SystemAndCommunicationsProtections,SystemAndInformationIntegrity",
          "policy.open-cluster-management.io/controls": "MutationAdvisor,VA",
          "policy.open-cluster-management.io/standards": "NIST,HIPAA",
          "seed-generation": "1",
        },
        "creationTimestamp": "2019-07-16T13:19:03Z",
        "finalizers": Array [
          "propagator.finalizer.calamari.ibm.com",
        ],
        "generation": 17,
        "name": "policy-role",
        "namespace": "calamari",
        "resourceVersion": "1513562",
        "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/calamari/policies/policy-role",
        "uid": "47dbc091-a7cc-11e9-8a41-005056a061f1",
      },
      "spec": Object {
        "namespaces": Object {
          "exclude": Array [
            "kube*",
          ],
          "include": Array [
            "default",
          ],
        },
        "remediationAction": "inform",
        "role-templates": Array [
          Object {
            "apiVersion": "roletemplate.calamari.ibm.com/v1alpha1",
            "complianceType": "musthave",
            "metadata": Object {
              "creationTimestamp": null,
              "name": "operator-role-policy",
            },
            "rules": Array [
              Object {
                "complianceType": "musthave",
                "policyRule": Object {
                  "apiGroups": Array [
                    "extensions",
                    "apps",
                  ],
                  "resources": Array [
                    "deployments",
                  ],
                  "verbs": Array [
                    "get",
                    "list",
                    "watch",
                    "create",
                    "delete",
                    "patch",
                  ],
                },
              },
            ],
            "selector": Object {
              "matchLabels": Object {
                "dev": "true",
              },
            },
            "status": Object {
              "Validity": Object {},
            },
          },
        ],
      },
      "status": Object {
        "placementBindings": Array [
          "binding-role",
        ],
        "placementPolicies": Array [
          "placement-role",
        ],
        "status": Object {
          "cluster1": Object {
            "aggregatePoliciesStatus": Object {
              "policy-role": Object {
                "compliant": "Compliant",
              },
            },
            "clustername": "cluster1",
            "compliant": "Compliant",
          },
          "clusterhub": Object {
            "aggregatePoliciesStatus": Object {
              "policy-role": Object {
                "compliant": "Compliant",
              },
            },
            "clustername": "clusterhub",
            "compliant": "Compliant",
          },
        },
      },
    },
    "remediation": "inform",
    "subItems": Array [
      Object {
        "items": Array [
          "binding-role",
        ],
        "name": "policy.pb",
      },
      Object {
        "items": Array [
          "placement-role",
        ],
        "name": "policy.pp",
      },
    ],
  },
]
`;

exports[`formatExpandablePolicies should correctly format 2`] = `Array []`;

exports[`formatExpandablePolicies should correctly format 3`] = `Array []`;

exports[`formatPoliciesToClustersTableData should correctly format 1`] = `Array []`;

exports[`formatPoliciesToClustersTableData should correctly format 2`] = `Array []`;

exports[`formatPoliciesToClustersTableData should correctly format 3`] = `Array []`;

exports[`formatPoliciesToClustersTableData should correctly format 4`] = `Array []`;
